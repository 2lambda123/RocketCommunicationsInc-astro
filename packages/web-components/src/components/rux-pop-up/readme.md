# rux-pop-up



<!-- Auto Generated Below -->


## Properties

| Property    | Attribute   | Description                                                           | Type                                                                                                                                                                 | Default      |
| ----------- | ----------- | --------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------ |
| `open`      | `open`      | determines if the pop up is open or closed                            | `boolean`                                                                                                                                                            | `false`      |
| `placement` | `placement` | the placement of the pop up relative to it's slotted trigger element. | `"bottom" \| "bottom-end" \| "bottom-start" \| "left" \| "left-end" \| "left-start" \| "right" \| "right-end" \| "right-start" \| "top" \| "top-end" \| "top-start"` | `'bottom'`   |
| `strategy`  | `strategy`  | The position strategy of the popup, either absolute or fixed.         | `"absolute" \| "fixed"`                                                                                                                                              | `'absolute'` |


## Events

| Event                  | Description                                                             | Type               |
| ---------------------- | ----------------------------------------------------------------------- | ------------------ |
| `ruxpopupmenuselected` | emits the value of the selected rux-menu-item inside of rux-pop-up-menu | `CustomEvent<any>` |


## Methods

### `hide() => Promise<false>`

Closes the pop up menu and returns false.

#### Returns

Type: `Promise<false>`



### `show() => Promise<true>`

Opens the pop up menu and returns true.

#### Returns

Type: `Promise<true>`




## Slots

| Slot          | Description                             |
| ------------- | --------------------------------------- |
| `"(default)"` | The contents for rux-pop-up-menu        |
| `"trigger"`   | The trigger element for rux-pop-up-menu |


## Shadow Parts

| Part                  | Description                                              |
| --------------------- | -------------------------------------------------------- |
| `"arrow"`             | the arrow pointing to the trigger of rux-pop-up-menu     |
| `"container"`         | the container of rux-pop-up-menu                         |
| `"popup-content"`     | the content that is shown when rux-pop-up-menu is opened |
| `"trigger-container"` | the container of the pop-up trigger                      |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
