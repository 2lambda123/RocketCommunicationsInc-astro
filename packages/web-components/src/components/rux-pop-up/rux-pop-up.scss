:host {
    display: block;
    width: min-content;
}

.hidden,
:host([hidden]) {
    display: none;
}
.hidden {
    display: none !important;
}

::slotted(*[slot='trigger']) {
    cursor: pointer;
}

.rux-popup__content {
    position: absolute;
    padding: var(--spacing-0); //no spacing, slotted content handles padding
    border-radius: var(--radius-base);
    background: var(--color-background-base-default);
    border: 1px solid var(--color-border-interactive-muted);
    transform-style: preserve-3d;
    z-index: 100; // overlays every component except for rux-dialog and rux-classification-marking
}

.rux-popup__trigger {
    display: contents;
}

.rux-popup__content--menu {
    padding: 0;
}
.rux-popup__arrow-left {
    border-right: 2px solid var(--color-border-interactive-muted);
}

.rux-popup__arrow-top {
    border-bottom: 2px solid var(--color-border-interactive-muted);
}
.rux-popup__arrow-right {
    border-left: 2px solid var(--color-border-interactive-muted);
}
.rux-popup__arrow-bottom {
    border-top: 2px solid var(--color-border-interactive-muted);
}

.rux-popup-arrow {
    position: absolute;
    background: var(--color-border-interactive-muted);
    width: var(--spacing-4); //16
    height: var(--spacing-4); //16
    transform: translateZ(-1px) rotate(45deg);
}
