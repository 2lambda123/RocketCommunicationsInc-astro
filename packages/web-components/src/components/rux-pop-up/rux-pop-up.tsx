import {
    Prop,
    Watch,
    Element,
    Component,
    Host,
    h,
    State,
    Event,
    EventEmitter,
    Method,
    Listen,
} from '@stencil/core'
import {
    computePosition,
    arrow,
    offset,
    flip,
    autoUpdate,
    autoPlacement,
} from '@floating-ui/dom'
import { ExtendedPlacement } from '../../common/commonTypes.module'

/**
 * @slot (default) - The contents for rux-pop-up
 * @slot trigger - The trigger element for rux-pop-up
 *
 * @part container - the container of rux-pop-up
 * @part trigger-container - the container of the pop-up trigger
 * @part popup-content - the content that is shown when rux-pop-up is opened
 * @part arrow - the arrow pointing to the trigger of rux-pop-up
 */

@Component({
    tag: 'rux-pop-up',
    styleUrl: 'rux-pop-up.scss',
    shadow: true,
})
export class RuxPopUp {
    private trigger!: HTMLElement
    private content!: HTMLElement
    private arrowEl!: HTMLElement
    private _positionerCleanup: ReturnType<typeof autoUpdate> | undefined

    @Element() el!: HTMLRuxPopUpElement

    /**
     * Determines if the pop up is open or closed
     */
    @Prop({ mutable: true, reflect: true }) open = false

    /**
     * The placement of the pop up relative to it's slotted trigger element. Defaults to auto.
     */
    @Prop() placement: ExtendedPlacement = 'auto'

    /**
     * Turns autoUpdate on or off which makes the pop-up move to stay in view based on scroll. Defaults to false.
     */
    @Prop({ reflect: true }) disableAutoUpdate: boolean = false

    /**
     * watches for trigger movements and replace the popup if movement is detected.
     */
    @Prop({ reflect: true }) enableAnimationFrame: boolean = false

    /**
     * The position strategy of the popup, either absolute or fixed.
     */
    @Prop() strategy: 'absolute' | 'fixed' = 'absolute'

    /**
     * When provided, will close the pop-up when a single selection is made.
     */
    @Prop({ attribute: 'close-on-select' }) closeOnSelect: boolean = false

    @State() arrowPosition?: string

    /**
     * Emits when the pop up has opened
     */
    @Event({ eventName: 'ruxpopupopened' })
    ruxPopUpOpened!: EventEmitter
    /**
     * Emits when the pop up has closed.
     */
    @Event({ eventName: 'ruxpopupclosed' })
    ruxPopUpClosed!: EventEmitter

    @Watch('open')
    handleOpen() {
        if (this.open) {
            this.content.style.display = 'block'
            this._startPositioner()
            this.ruxPopUpOpened.emit()
            window.addEventListener('mousedown', (e: MouseEvent) =>
                this._handleOutsideClick(e)
            )
        } else {
            this.content.style.display = ''
            this._stopPositioner()
            this.ruxPopUpClosed.emit()
            window.removeEventListener('mousedown', (e: MouseEvent) =>
                this._handleOutsideClick(e)
            )
        }
    }

    /**
     * Opens the pop up and returns true.
     */
    @Method()
    async show() {
        if (this.open) {
            return this.open
        } else this.open = true
        return this.open
    }

    /**
     * Closes the pop up and returns false.
     */
    @Method()
    async hide() {
        if (!this.open) {
            return this.open
        } else this.open = false
        return this.open
    }

    connectedCallback() {
        this._handleTriggerClick = this._handleTriggerClick.bind(this)
        this._handleTriggerKeyPress = this._handleTriggerKeyPress.bind(this)
        this._handleOutsideClick = this._handleOutsideClick.bind(this)
        this._setTriggerTabIndex = this._setTriggerTabIndex.bind(this)
    }

    componentWillLoad() {
        this._setTriggerTabIndex()
    }

    componentDidRender() {
        if (this.open) {
            this.handleOpen()
        }
    }

    private async _handleTriggerClick(event: MouseEvent) {
        //excludes synthetic keyboard events such as Enter on a button behaving as a click
        if (event.detail > 0) {
            this.open = !this.open
        }
    }

    //part of focus + accessible keyboard events
    private async _handleTriggerKeyPress(event: KeyboardEvent) {
        if (event.key === 'Enter') {
            this.open = !this.open
        }
    }

    private _position() {
        if (!this.open || !this.triggerSlot || !this.content) {
            return
        }
        const placementCheck = () => {
            // disable auto update = false, placement is anything
            if (!this.disableAutoUpdate) {
                return [
                    offset(12),
                    this.placement === 'auto'
                        ? autoPlacement({ alignment: 'start' })
                        : flip(),
                    arrow({ element: this.arrowEl }),
                ]
            }
            // disableAutoUpdate = true, placement=auto
            else if (this.placement === 'auto') {
                /* we need to set a starting placement for the user because 'auto' is only understood by floating-ui as something means,
            'pick whatever side fits best', and in this case we don't want any automatic setting of placements */
                this.placement = 'top'
                return [offset(12), arrow({ element: this.arrowEl })]
                // disableAutoUpdate = true, placement != auto
            } else {
                return [offset(12), arrow({ element: this.arrowEl })]
            }
        }
        computePosition(this.triggerSlot, this.content, {
            //@ts-ignore
            placement: this.placement,
            strategy: this.strategy,
            middleware: placementCheck(),
        }).then(({ x, y, placement, middlewareData }) => {
            Object.assign(this.content.style, {
                left: `${x}px`,
                top: `${y}px`,
                position: `${this.strategy}`,
            })

            //@ts-ignore
            const { x: arrowX, y: arrowY } = middlewareData.arrow

            const staticSide = {
                top: 'bottom',
                right: 'left',
                bottom: 'top',
                left: 'right',
            }[placement.split('-')[0]]
            Object.assign(this.arrowEl!.style, {
                left: arrowX != null ? `${arrowX}px` : '',
                top: arrowY != null ? `${arrowY}px` : '',
                right: '',
                bottom: '',
                //@ts-ignore
                [staticSide]: '-6px',
            })
        })
        this._setArrowPosition()
    }

    //set a tabindex if the trigger element does not have one, necessary for items that can not natively receive focus
    private _setTriggerTabIndex() {
        const triggerEl = this.el.querySelector(
            `[slot='trigger']`
        ) as HTMLElement
        triggerEl.hasAttribute('tabindex')
            ? null
            : triggerEl.setAttribute('tabindex', '0')
    }

    private _startPositioner() {
        this._stopPositioner()
        if (this.open) {
            this._position()
            this._positionerCleanup = autoUpdate(
                this.triggerSlot,
                this.content,
                this._position.bind(this),
                { animationFrame: this.enableAnimationFrame }
            )
        }
    }

    private async _setArrowPosition() {
        const arrowPos = await this._determineArrowPosition()
        if (this.arrowPosition === arrowPos) {
            return
        } else this.arrowPosition = arrowPos
    }

    private async _determineArrowPosition() {
        if (!this.open) {
            return
        }

        const triggerElRect = await this.getTriggerRect()
        const arrowDivRect = await this.getArrowRect()

        if (
            triggerElRect.bottom > arrowDivRect.bottom &&
            triggerElRect.top < arrowDivRect.top
        ) {
            if (triggerElRect.right > arrowDivRect.right) {
                return 'left'
            } else {
                return 'right'
            }
        } else {
            if (triggerElRect.bottom > arrowDivRect.bottom) {
                return 'top'
            } else {
                return 'bottom'
            }
        }
    }
    private async getTriggerRect() {
        return this.triggerSlot.getBoundingClientRect()
    }
    private async getArrowRect() {
        return this.arrowEl?.getBoundingClientRect()
    }

    private _stopPositioner() {
        if (this._positionerCleanup) {
            this._positionerCleanup()
            this._positionerCleanup = undefined
        }
    }

    private _handleOutsideClick(e: MouseEvent) {
        const menuClick = e.composedPath().includes(this.contentSlot)
        const triggerClick = e.composedPath().includes(this.triggerSlot)
        const popUpClick = e.composedPath().includes(this.el)
        if (!menuClick && !triggerClick && !popUpClick) {
            this.open = false
        }
    }

    @Listen('ruxmenuselected')
    handleSelection() {
        if (this.closeOnSelect) {
            this.open = false
        }
    }

    get contentSlot() {
        return this.content
            ?.querySelector('slot')!
            .assignedElements({ flatten: true })[0]
    }

    get triggerSlot() {
        return this.trigger
            ?.querySelector('slot')!
            .assignedElements({ flatten: true })[0]
    }

    get hasMenu(): boolean {
        //@ts-ignore
        return !!this.content
            ?.querySelector('slot')
            .assignedElements({ flatten: true })
            .filter(
                (el) => el.tagName.toLowerCase() === 'rux-menu'
            )[0] as HTMLRuxMenuElement
    }

    render() {
        return (
            <Host>
                <div class="rux-popup" part="container">
                    <div
                        onClick={this._handleTriggerClick}
                        onKeyPress={this._handleTriggerKeyPress}
                        class="rux-popup__trigger"
                        ref={(el) => (this.trigger = el!)}
                        part="trigger-container"
                    >
                        <slot name="trigger"></slot>
                    </div>

                    <div
                        class={{
                            'rux-popup__arrow-left':
                                this.arrowPosition === 'left',
                            'rux-popup__arrow-top':
                                this.arrowPosition === 'top',
                            'rux-popup__arrow-right':
                                this.arrowPosition === 'right',
                            'rux-popup__arrow-bottom':
                                this.arrowPosition === 'bottom',
                            'rux-popup__content': true,
                            'rux-popup__content--menu': this.hasMenu,
                            hidden: this.open === false,
                        }}
                        aria-hidden={this.open ? 'false' : 'true'}
                        part="popup-content"
                        ref={(el) => (this.content = el!)}
                    >
                        <div
                            class="rux-popup-arrow"
                            ref={(el) => (this.arrowEl = el!)}
                            part="arrow"
                        ></div>

                        <slot></slot>
                    </div>
                </div>
            </Host>
        )
    }
}
