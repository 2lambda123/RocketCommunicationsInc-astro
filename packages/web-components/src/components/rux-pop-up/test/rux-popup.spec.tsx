import { newSpecPage } from '@stencil/core/testing'
import { RuxPopUp } from '../rux-pop-up'

describe('rux-pop-up', () => {
    it('renders', async () => {
        const page = await newSpecPage({
            components: [RuxPopUp],
            html: `<rux-pop-up></rux-pop-up>`,
        })
        expect(page.root).toMatchSnapshot()
    })
    it('closes the pop up with the hide method', () => {
        const popup = new RuxPopUp()
        popup.open = true
        expect(popup.open).toBe(true)
        popup.hide()
        expect(popup.open).toBe(false)
    })
    it('opens the pop up with the show method', () => {
        const popup = new RuxPopUp()
        expect(popup.open).toBe(false)
        popup.show()
        expect(popup.open).toBe(true)
    })
})
