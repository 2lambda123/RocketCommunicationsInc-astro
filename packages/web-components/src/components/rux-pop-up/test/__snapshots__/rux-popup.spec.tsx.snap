// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rux-pop-up-menu renders 1`] = `
<rux-pop-up-menu aria-hidden="true">
  <mock:shadow-root>
    <div class="rux-popup" part="container">
      <div class="rux-popup__trigger" part="trigger-container">
        <slot name="trigger"></slot>
      </div>
      <div class="hidden rux-popup__content" part="popup-content">
        <div class="rux-popup-arrow" part="arrow"></div>
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
</rux-pop-up-menu>
`;
